### ISSUE TITLE
Medusa import with encrypt -- "panic: runtime error: index out of range [0] with length 0"

### ISSUE THREAD

**Issue by cebaldauff** [jonasvinther/medusa#84]
Hey there, so I am importing several sets of encrypted secrets into a private vault instance. About half of them import just fine, the other half fail with the following message:
`panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/jonasvinther/medusa/pkg/encrypt.Decrypt(0x7ffd3766a689, 0x19, 0x7ffd3766a5fd, 0x30, 0xc, 0xc0001296d7, 0x1, 0x0)
	/vault/medusa/pkg/encrypt/encrypt.go:46 +0x41d
github.com/jonasvinther/medusa/cmd.glob..func4(0xc62a20, 0xc000184300, 0x2, 0x8, 0x0, 0x0)
	/vault/medusa/cmd/import.go:51 +0x4a5
github.com/spf13/cobra.(*Command).execute(0xc62a20, 0xc000184280, 0x8, 0x8, 0xc62a20, 0xc000184280)
	/root/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc62240, 0xc000000180, 0xc00018df78, 0x407d05)
	/root/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/root/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
github.com/jonasvinther/medusa/cmd.Execute(...)
	/vault/medusa/cmd/cmd.go:59
main.main()
	/vault/medusa/main.go:10 +0x32`

All sets were successfully exported from vault using "medusa export" with --encrypt. The difference between the sets that successfully decrypt and import and the ones that do not is that the failing sets are larger, with more sets of secrets.

Thanks for any help you can provide.

**Comment by cebaldauff** [comment]
Unrelated but I would also like to be able to import secrets not under "data/". Is there any way to change that?

**Comment by cebaldauff** [comment]
For whatever reason, this extra "data" field is being added on to the secrets' path: 
![image](https://user-images.githubusercontent.com/50680435/168380412-98882fe9-314e-45c7-8960-439e99d7b9c0.png)

**Comment by jonasvinther** [comment]
Hi @cebaldauff 
Please check if you are moving secrets between kv1 and kv2 secret engines. The `data` prefix is something that is specific for kv2. So I think that you maybe are exporting from a kv2 into a kv1 secret engine.

**Comment by cebaldauff** [comment]
Thank you so much, @jonasvinther, that was it exactly! I'm still unable to decrypt when a file is large enough. Is there any easy fix for that issue?

**Comment by jonasvinther** [comment]
I will have to look into the issue with decrypting large files. Can you say anything about how large your files/secrets are? Are you storing passwords or larger amount of data in a secret? Like ssh keys or something like that.

**Comment by cebaldauff** [comment]
Some of the secret sets that are being imported and failing on decrypt include either 60+ sets of secrets, or they include k8s  configs.

**Comment by cebaldauff** [comment]
Also seeing this when reading in a yaml which is causing a set of secrets to fail:
`yaml: control characters are not allowed
Error: yaml: control characters are not allowed`

Found this on stackoverflow, could be this problem: https://stackoverflow.com/questions/33717799/go-yaml-control-characters-are-not-allowed-error
And here is another github project that solved the issue: https://github.com/projectdiscovery/nuclei/issues/437

**Comment by cebaldauff** [comment]
> Also seeing this when reading in a yaml which is causing a set of secrets to fail: `yaml: control characters are not allowed Error: yaml: control characters are not allowed`
> 
> Found this on stackoverflow, could be this problem: https://stackoverflow.com/questions/33717799/go-yaml-control-characters-are-not-allowed-error And here is another github project that solved the issue: [projectdiscovery/nuclei#437](https://github.com/projectdiscovery/nuclei/issues/437)

This is different from the decrypt problem, I can open a separate ticket for this if you would like.

**Comment by cebaldauff** [comment]
Hey @jonasvinther please let me know if there is anything I can do to assist on this issue. Thanks!

**Comment by jonasvinther** [comment]
Hi @cebaldauff 
Can you provide a file that contains some dummy secrets in clear text? Then I can import that into Vault and try and do the encrypted export+import reproduce the error. That would be a lot of help.
And great finding you did regarding the other problem, I can have a look at that as well. Again it's easier for me if you can provide some dummy file that so that I can reproduce the error.

**Comment by motilevy** [comment]
To add to this bug. 

I have a job running medusa backups ( exports ) for over a year. The medusa version did not change ( I have it in a docker container ) I think the issue is related to size/amount of items. 
The following 2 backups are 1 hour apart, using the same docker image and I am able to restore the 62kb file but not the 82kb file. 

<img width="907" alt="image" src="https://user-images.githubusercontent.com/3513686/178715380-feb17cc8-cd24-4bbb-81e9-2c890cfe180d.png">

**Comment by jmhbnz** [comment]
Came across this issue today attempting to import encryped secrets into a kv2 secret engine.

Here was the output:

```bash
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/jonasvinther/medusa/pkg/encrypt.Decrypt({0x7ffe355d640f, 0x15}, {0x7ffe355d638c?, 0x40?})
        /home/runner/work/medusa/medusa/pkg/encrypt/encrypt.go:46 +0x2cd
github.com/jonasvinther/medusa/cmd.glob..func4(0xc20640?, {0xc0000c6720?, 0x6?, 0x6?})
        /home/runner/work/medusa/medusa/cmd/import.go:51 +0x327
github.com/spf13/cobra.(*Command).execute(0xc20640, {0xc0000c66c0, 0x6, 0x6})
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0xc1fe60)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x39c
github.com/spf13/cobra.(*Command).Execute(...)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
github.com/jonasvinther/medusa/cmd.Execute(...)
        /home/runner/work/medusa/medusa/cmd/cmd.go:59
main.main()
        /home/runner/work/medusa/medusa/main.go:10 +0x25
```

It was for two decent size files, without encryption the yaml is ~690 lines long and includes a number of more complex secrets like private keys etc.  

The following files were exported via medusa with encryption:
```bash
-rw-r--r--. 1 blairj domain users 13974 Jul 26 08:56 common.txt
-rw-r--r--. 1 blairj domain users 43870 Jul 26 08:56 dev.txt
-rw-r--r--. 1 blairj domain users  1506 Jul 26 08:57 pentest.txt
-rw-r--r--. 1 blairj domain users 88638 Jul 26 08:57 prod.txt
-rw-r--r--. 1 blairj domain users  2854 Jul 26 08:57 staging.txt
-rw-r--r--. 1 blairj domain users  1254 Jul 26 08:57 test-automation.txt
-rw-r--r--. 1 blairj domain users 85330 Jul 26 08:57 test.txt
```

The two that failed import were the large `prod.txt` and `test.txt` files however interestingly the `dev.txt` file contains similar complex secrets just at less volume and that imported as expected.

**Comment by infa-amohanda** [comment]
We are also facing issues with large size files, any workaround for this ?

**Comment by jonasvinther** [comment]
Hello awesome people. I have finally gotten some time to look into this issue. The issue has been identified and fixed in #111. Thank you all for your patience 🙏🏼
